### ISSUE TITLE
Modular encoder fails for tall/wide images

### ISSUE THREAD

**Issue by eustas** [libjxl/libjxl#3937]
```
convert -size 15x167941 xc: tall.png
# if fails, update /etc/ImageMagick-6/policy.xml

./build/tools/cjxl ./tall.png ./tall.jxl
# OK

./build/tools/cjxl -m 1 ./tall.png ./tall.jxl
# Fails
```

Error:
```
JPEG XL encoder v0.12.0 538c140d [AVX2,SSE4,SSE2]
Encoding [Modular, d1.000, effort: 7]
lib/jxl/modular/encoding/encoding.cc:523: JXL_FAILURE: Inconsistent transforms
lib/jxl/enc_modular.cc:1036: JXL_RETURN_IF_ERROR code=1: ValidateChannelDimensions(gi, stream_options_[0])
lib/jxl/enc_frame.cc:1612: JXL_RETURN_IF_ERROR code=1: enc_modular.ComputeEncodingData( frame_header, metadata->m, &color, extra_channels, group_rect, frame_dim, frame_area_rect, &enc_state, cms, pool, aux_out, cparams.modular_mode)
lib/jxl/enc_frame.cc:2125: JXL_RETURN_IF_ERROR code=1: ComputeEncodingData( cparams, frame_info, metadata, frame_data, jpeg_data.get(), 0, 0, frame_data.xsize, frame_data.ysize, cms, pool, frame_header, enc_modular, enc_state, &group_codes, aux_out)
lib/jxl/encode.cc:999: Failed to encode frame
JxlEncoderProcessOutput failed.
EncodeImageJXL() failed.
```

**Comment by eustas** [comment]
Disabling this check causes considerable diff in "streaming_fuzzer" (-> failure).

**Comment by eustas** [comment]
@jonsneyers could you take a look at `ValidateChannelDimensions` / `DefaultSqueezeParameters`. Likely one is over-restrictive or the other is not taking into account group dim...

**Comment by jonnyawsom3** [comment]
Should be fixed by #3942 (I think).

**Comment by eustas** [comment]
Re-open; a case found where `ValidateChannelDimensions` fails.

**Comment by eustas** [comment]
```
Image size: 57601 X 1, d=0.000000
JXL_ENC_FRAME_SETTING_EFFORT = 8
JXL_ENC_FRAME_SETTING_DECODING_SPEED = 3
JXL_ENC_FRAME_SETTING_NOISE = -1
JXL_ENC_FRAME_SETTING_DOTS = -1
JXL_ENC_FRAME_SETTING_PATCHES = -1
JXL_ENC_FRAME_SETTING_EPF = -1
JXL_ENC_FRAME_SETTING_GABORISH = 0
JXL_ENC_FRAME_SETTING_MODULAR = -1
JXL_ENC_FRAME_SETTING_KEEP_INVISIBLE = 0
JXL_ENC_FRAME_SETTING_RESPONSIVE = 1
JXL_ENC_FRAME_SETTING_PROGRESSIVE_AC = 1
JXL_ENC_FRAME_SETTING_QPROGRESSIVE_AC = -1
JXL_ENC_FRAME_SETTING_PROGRESSIVE_DC = 0
JXL_ENC_FRAME_SETTING_PALETTE_COLORS = 51
JXL_ENC_FRAME_SETTING_LOSSY_PALETTE = 1
JXL_ENC_FRAME_SETTING_COLOR_TRANSFORM = 2
JXL_ENC_FRAME_SETTING_MODULAR_COLOR_SPACE = 27
JXL_ENC_FRAME_SETTING_MODULAR_GROUP_SIZE = -1
JXL_ENC_FRAME_SETTING_MODULAR_PREDICTOR = -1
JXL_ENC_FRAME_SETTING_MODULAR_NB_PREV_CHANNELS = -1
```

Cause `lib/jxl/modular/encoding/encoding.cc:529: JXL_FAILURE: Inconsistent transforms`

**Comment by eustas** [comment]
The largest non-failing `xsize` is `16448 == 16384 + 64`.

**Comment by eustas** [comment]
I propose to validate `DefaultSqueezeParams` output.

When decoding - reject stream even earlier if `max(count(horizontal==true),count(horizontal==false)) > 7 + group_size_shift`.
Also add remarks to spec in `ModularGroup` / `ModularLfGroup` that after shifting group dim should not be less than 1.

When encoding - if the same check fails, then truncate params until check passes; encode those params (override default).

That way it won't break decoding existing images (just detect earlier invalid streams), does not require spec changes (modulo clarification remarks), allows encoding images larger than 16k-128k (depending on group shift).

@jonsneyers WDYT

**Comment by jonsneyers** [comment]
Could you remind me what is causing the encode to fail?
I think at some point we did allow modular channel dimensions to have 0 width or height (they are just dummy channels that require no bits to encode), and then there was no problem with DefaultSqueezeParams.

But then I think at some point fuzzing found cases where it could cause issues (the shift became too large in some cases when doing multiple default squeezes, or something? I don't remember exactly), and then we made things stricter and stopped allowing channels with 0 width or height. This was an issue that never occurred in 'normal' bitstreams produced by the encoder.

I'd prefer if we could allow degenerate channels again so DefaultSqueezeParams always produces something valid (even though it may create zero-dimension channels in case the image has a very extreme aspect ratio), and fix the fuzzing problem in some other way.

**Comment by jonnyawsom3** [comment]
H.2 is the only part in the spec I can find that mentions 0 width or height channels.
> The decoder then starts an entropy-coded stream ([C.1](https://github.com/libjxl/libjxl/issues/3937#ans)) and decodes the data for each channel (in ascending order of index) as specified in [H.3](https://github.com/libjxl/libjxl/issues/3937#channel-decoding), skipping any channels having width or height zero.

**Comment by eustas** [comment]
`group_dim >> shift` becomes `0`. And that breaks all the group arithmetic, e.g. all groups coincide (because dx/dy in grid becomes 0).

**Comment by jonsneyers** [comment]
But it's ok if all groups coincide, it's a channel with zero pixels anyway so nothing will happen with it anyway. I am sure there was something that broke somewhere, but I think it's better to allow empty channels and only prevent the thing that actually caused an issue in the fuzzer. I don't remember what it was, maybe shift itself getting too large when you make a crazy chain of multiple squeezes?

**Comment by eustas** [comment]
Here we have another problem, different from 0-dim channel.

0-dim group is worse because it is not clear how pixels are distributed (and there are pixels, because 0 is just rounding for what would be `2^-n`).

Quote from spec: `then a channel ... is added, ... where the group dimensions and the x,y offsets are right-shifted by hshift (for x and width) and vshift (for y and height).`

**Comment by jonsneyers** [comment]
Right. So what is in the spec becomes ill-defined in case `(group_dim >> shift) == 0` and at the same time `(width >> hshift) > 0` (or the same for `vshift` and `height`).
I understand better now what's the issue here. For sufficiently large dimensions, DefaultSqueezeParameters will indeed go into values of hshift/vshift that can cause such trouble, since it keeps increasing the hshift/vshift until you get something that is at most 8x8.

OK, I agree that the approach you sketched in https://github.com/libjxl/libjxl/issues/3937#issuecomment-3000241430 is a good one.

Make sure that the decode-side reject criterion is not too strict: remember that the smaller channels (the ones that fit in a single group) are encoded in GlobalModular which does not have this problem, but it's specifically the extreme aspect ratio that causes an issue because then you get channels that are too large in one dimension to fit in GlobalModular but still shifted too much to cause an issue in the group-based sections.

Adding a spec clarification that ill-defined stuff does not happen is OK, we can do that in the next edition of 18181-1.
Making sure not to let the encoder do that is a good idea. Using non-default parameters should not really be an issue. The signaling overhead of that will be a few bytes, which should be irrelevant for images with such large dimensions (even if they have an extreme aspect ratio so the total number of pixels is not super huge).

It would perhaps conceptually be nicer to change `DefaultSqueezeParameters` to never produce parameters that will cause ill-defined stuff later on, but that's an actual spec change and will break existing decoders. So it's basically too late for that now.

So yes, the approach you sketched is a good one.

**Comment by eustas** [comment]
To stay on the safe side, I'll currently leave decoder as-is. It still does `ValidateChannelDimensions` to catch the problem (after MetaApply, so we only pay with memory allocations, and only in case of "invalid" stream).

Testing encoder PR. Will publish soon.
